src/lib.rs

```diff
@@ -233,13 +279,41 @@ impl<'a, T> ReadLock<'a, T> {
 
     /// Pin an iterated item with a newly created `Pointer`.
     pub fn pin(&self, item: &ReadItem<'a, T>) -> Pointer<T> {
-        self.pending.lock().unwrap().add_ref.push(item.index);
+        let mut pending = self.pending.lock().unwrap();
+        pending.add_ref.push(item.index);
         Pointer {
             index: item.index,
+            epoch: pending.epoch[item.index],
             target: self.storage.clone(),
             pending: self.pending.clone(),
         }
     }
+
+    /// Upgrades the `WeakPointer` to an `Pointer`, if possible.
+    /// Returns `Err` if the strong count has reached zero and the inner value was destroyed.
+    pub fn upgrade(&self, ptr: &WeakPointer<T>) -> Result<Pointer<T>, UpgradeErr> {
```

Comment:
we'll need to seriously make this thing nice
first off, when `epoch` is bumped on deletion, this code no longer needs to check for the refcount == 0
secondly, now that we don't need `guard`, we don't actually need `&self` to be `ReadLock`/`WriteLock` either!
so we can make `upgrade` just a method of `WeakPointer`, as you initially wanted, but without the danger of dead-locking the `Storage`.